packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,256 @@
+      new AsyncSnapshot<T>(ConnectionState.active, data, null);
+
+  @override
+  AsyncSnapshot<T> onError(AsyncSnapshot<T> current, dynamic error) =>
+      new AsyncSnapshot<T>(ConnectionState.active, null, error);
+
+  @override
+  AsyncSnapshot<T> onDone(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(ConnectionState.done, current.data, current.error);
+
+  @override
+  AsyncSnapshot<T> onDisconnecting(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(
+          ConnectionState.none, current.data, current.error);
+
+  @override
+  Widget build(BuildContext context, AsyncSnapshot<T> currentSummary) =>
+      builder(context, currentSummary);
+}
+
+/// Base class for Widgets that build themselves based on interaction with
+/// a specified [stream].
+///
+/// The widget's state maintains a summary of the interaction so far. The
+/// summary is defined by sub-classes.
+///
+/// [T] type of stream events.
+/// [S] type of summary.
+abstract class StreamFold<T, S> extends StatefulWidget {
+  final Stream<T> stream;
+
+  StreamFold({Key key, this.stream}) : super(key: key);
+
+  /// Returns the initial summary.
+  S initial();
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are now connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S onConnecting(S current) => current;
+
+  /// Returns an updated version of the [current] summary following an event.
+  S onData(S current, T data);
+
+  /// Returns an updated version of the [current] summary following an error.
+  ///
+  /// The default implementation returns [current] as is.
+  S onError(S current, dynamic error) => current;
+
+  /// Returns an updated version of the [current] summary following stream
+  /// termination.
+  ///
+  /// The default implementation returns [current] as is.
+  S onDone(S current) => current;
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are no longer connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S onDisconnecting(S current) => current;
+
+  /// Returns a Widget based on the [currentSummary].
+  Widget build(BuildContext context, S currentSummary);
+
+  @override
+  State<StreamFold<T, S>> createState() => new _StreamFoldState<T, S>();
+}
+
+class _StreamFoldState<T, S> extends State<StreamFold<T, S>> {
+  StreamSubscription<T> _subscription;
+  S _summary;
+
+  @override
+  void initState() {
+    super.initState();
+    _summary = config.initial();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(StreamFold<T, S> oldConfig) {
+    if (!identical(oldConfig.stream, config.stream)) {
+      _unsubscribe();
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) => config.build(context, _summary);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    if (config.stream != null) {
```

Comment:
since this is only ever called from initState or didUpdateConfig, there's no need to call setState in this method.